tensorflow/python/keras/layers/wrappers.py

```diff
@@ -148,56 +150,73 @@ def _get_shape_tuple(self, init_tuple, tensor, start_idx, int_shape=None):
         of the output shape
 
     Returns:
-      The new int_shape with the first part from init_tuple
-      and the last part from either `int_shape` (if provided)
-      or `tensor.shape`, where every `None` is replaced by
-      the corresponding dimension from `tf.shape(tensor)`.
+      A tensor with the new int_shape with the first part from
+      init_tuple and the last part from either `int_shape`
+      (if provided) or `tensor.shape`, where every `None` is
+      replaced by the corresponding dimension from `tf.shape(tensor)`.
     """
     # replace all None in int_shape by K.shape
     if int_shape is None:
       int_shape = K.int_shape(tensor)[start_idx:]
     if not any(not s for s in int_shape):
-      return init_tuple + tuple(int_shape)
+      return convert_to_tensor(init_tuple + tuple(int_shape))
     shape = K.shape(tensor)
     int_shape = list(int_shape)
     for i, s in enumerate(int_shape):
       if not s:
         int_shape[i] = shape[start_idx + i]
-    return init_tuple + tuple(int_shape)
+    return convert_to_tensor(init_tuple + tuple(int_shape))
 
   def build(self, input_shape):
-    input_shape = tensor_shape.TensorShape(input_shape).as_list()
-    if len(input_shape) < 3:
+    input_shape = tf_utils.convert_shapes(input_shape, to_tuples=False)
+    input_dims = generic_utils.to_list(nest.map_structure(lambda x: x.ndims, input_shape))
+    if all(dim < 3 for dim in input_dims):
       raise ValueError(
           '`TimeDistributed` Layer should be passed an `input_shape ` '
           'with at least 3 dimensions, received: ' + str(input_shape))
     # Don't enforce the batch or time dimension.
-    self.input_spec = InputSpec(shape=[None, None] + input_shape[2:])
-    child_input_shape = [input_shape[0]] + input_shape[2:]
+    self.input_spec = nest.map_structure(lambda x: InputSpec(shape=[None, None] + x.as_list()[2:]),
+                                         input_shape)
+    child_input_shape = nest.map_structure(
+        lambda x: tensor_shape.TensorShape([x.as_list()[0]] + x.as_list()[2:]),
+        input_shape)
+    child_input_shape = tf_utils.convert_shapes(child_input_shape)
     super(TimeDistributed, self).build(tuple(child_input_shape))
     self.built = True
 
   def compute_output_shape(self, input_shape):
-    input_shape = tensor_shape.TensorShape(input_shape).as_list()
-    child_input_shape = tensor_shape.TensorShape([input_shape[0]] +
-                                                 input_shape[2:])
+    input_shape = tf_utils.convert_shapes(input_shape, to_tuples=False)
+    def remove_timesteps(dims):
```

Comment:
It seems that this function can be used above for child_input_shape as well, can we move it to python module level?